src/dyncfg/src/lib.rs

```diff
@@ -234,44 +221,118 @@ impl ConfigEntry {
+/// shared between configuration writers and readers.
+// TODO(cfg): Consider moving these Arcs to be a single one around the map in
+// `ConfigSet` instead. That would mean less pointer-chasing in the common
+// case, but would remove the possibility of amortizing the name lookup via
+// [Config::get_from_shared].
+#[derive(Clone, Debug)]
+enum ConfigValAtomic {
     Bool(Arc<AtomicBool>),
-    /// A `u32` shared value.
     U32(Arc<AtomicU32>),
-    /// A `usize` shared value.
     Usize(Arc<AtomicUsize>),
-    /// An `Option<usize>` shared value.
     OptUsize(Arc<RwLock<Option<usize>>>),
-    /// A `String` shared value.
     String(Arc<RwLock<String>>),
-    /// A 'Duration' shared value.
     Duration(Arc<RwLock<Duration>>),
 }
 
+impl From<ConfigVal> for ConfigValAtomic {
+    fn from(val: ConfigVal) -> ConfigValAtomic {
+        match val {
+            ConfigVal::Bool(x) => ConfigValAtomic::Bool(Arc::new(AtomicBool::new(x))),
+            ConfigVal::U32(x) => ConfigValAtomic::U32(Arc::new(AtomicU32::new(x))),
+            ConfigVal::Usize(x) => ConfigValAtomic::Usize(Arc::new(AtomicUsize::new(x))),
+            ConfigVal::OptUsize(x) => ConfigValAtomic::OptUsize(Arc::new(RwLock::new(x))),
+            ConfigVal::String(x) => ConfigValAtomic::String(Arc::new(RwLock::new(x))),
+            ConfigVal::Duration(x) => ConfigValAtomic::Duration(Arc::new(RwLock::new(x))),
+        }
+    }
+}
+
+impl ConfigValAtomic {
+    fn load(&self) -> ConfigVal {
+        match self {
+            ConfigValAtomic::Bool(x) => ConfigVal::Bool(x.load(SeqCst)),
+            ConfigValAtomic::U32(x) => ConfigVal::U32(x.load(SeqCst)),
+            ConfigValAtomic::Usize(x) => ConfigVal::Usize(x.load(SeqCst)),
+            ConfigValAtomic::OptUsize(x) => ConfigVal::OptUsize(*x.read().expect("lock poisoned")),
+            ConfigValAtomic::String(x) => {
+                ConfigVal::String(x.read().expect("lock poisoned").clone())
+            }
+            ConfigValAtomic::Duration(x) => ConfigVal::Duration(*x.read().expect("lock poisoned")),
+        }
+    }
+
+    fn store(&self, val: ConfigVal) {
+        match (self, val) {
+            (ConfigValAtomic::Bool(x), ConfigVal::Bool(val)) => x.store(val, SeqCst),
+            (ConfigValAtomic::U32(x), ConfigVal::U32(val)) => x.store(val, SeqCst),
+            (ConfigValAtomic::Usize(x), ConfigVal::Usize(val)) => x.store(val, SeqCst),
+            (ConfigValAtomic::OptUsize(x), ConfigVal::OptUsize(val)) => {
+                *x.write().expect("lock poisoned") = val
+            }
+            (ConfigValAtomic::String(x), ConfigVal::String(val)) => {
+                *x.write().expect("lock poisoned") = val
+            }
+            (ConfigValAtomic::Duration(x), ConfigVal::Duration(val)) => {
+                *x.write().expect("lock poisoned") = val
+            }
+            (_, val) => panic!("attempted to store {val:?} value in {self:?} parameter"),
```

Comment:
Oh lol. I mentioned in slack that I loved us getting rid of the "Restructure this match so that it's harder to miss when adding a new `ConfigType`." TODO, but I think we kept the problem and just lost the TODO :)

At least the single way to set/get things means we have an easier time making sure it's covered by unit tests, so I still think this is an improvement

Reply:
Oh, it's easy enough to make this exhaustive. Done!